pubsub/pubsub_test.go

```diff
@@ -475,3 +476,127 @@ func TestShutdownsDoNotLeakGoroutines(t *testing.T) {
 		time.Sleep(time.Millisecond)
 	}
 }
+
+var (
+	testOpenOnce sync.Once
+	testOpenGot  *url.URL
+)
+
+func TestURLMux(t *testing.T) {
+	ctx := context.Background()
+	var gotTopic, gotSub *url.URL
+
+	mux := new(pubsub.URLMux)
+	// Register scheme foo to always return nil. Sets got as a side effect
+	mux.RegisterTopic("foo", topicURLOpenFunc(func(_ context.Context, u *url.URL) (*pubsub.Topic, error) {
+		gotTopic = u
+		return nil, nil
+	}))
+	mux.RegisterSubscription("foo", subscriptionURLOpenFunc(func(_ context.Context, u *url.URL) (*pubsub.Subscription, error) {
+		gotSub = u
+		return nil, nil
+	}))
+	// Register scheme err to always return an error.
+	mux.RegisterTopic("err", topicURLOpenFunc(func(_ context.Context, u *url.URL) (*pubsub.Topic, error) {
+		return nil, errors.New("fail")
+	}))
+	mux.RegisterSubscription("err", subscriptionURLOpenFunc(func(_ context.Context, u *url.URL) (*pubsub.Subscription, error) {
+		return nil, errors.New("fail")
+	}))
+
+	for _, tc := range []struct {
+		name    string
+		url     string
+		wantErr bool
+	}{
+		{
+			name:    "empty URL",
+			wantErr: true,
+		},
+		{
+			name:    "invalid URL",
+			url:     ":foo",
+			wantErr: true,
+		},
+		{
+			name:    "invalid URL no scheme",
+			url:     "foo",
+			wantErr: true,
+		},
+		{
+			name:    "unregistered scheme",
+			url:     "bar://myps",
+			wantErr: true,
+		},
+		{
+			name:    "func returns error",
+			url:     "err://myps",
+			wantErr: true,
+		},
+		{
+			name: "no query options",
+			url:  "foo://myps",
+		},
+		{
+			name: "empty query options",
+			url:  "foo://myps?",
+		},
+		{
+			name: "query options",
+			url:  "foo://myps?aAa=bBb&cCc=dDd",
+		},
+		{
+			name: "multiple query options",
+			url:  "foo://myps?x=a&x=b&x=c",
+		},
+		{
+			name: "fancy ps name",
+			url:  "foo:///foo/bar/baz",
+		},
+	} {
+		t.Run("topic: "+tc.name, func(t *testing.T) {
+			_, gotErr := mux.OpenTopic(ctx, tc.url)
+			if (gotErr != nil) != tc.wantErr {
+				t.Fatalf("got err %v, want error %v", gotErr, tc.wantErr)
+			}
+			if gotErr != nil {
+				return
+			}
+			want, err := url.Parse(tc.url)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if diff := cmp.Diff(gotTopic, want); diff != "" {
+				t.Errorf("got\n%v\nwant\n%v\ndiff\n%s", gotTopic, want, diff)
+			}
+		})
+		t.Run("subscription: "+tc.name, func(t *testing.T) {
+			_, gotErr := mux.OpenSubscription(ctx, tc.url)
+			if (gotErr != nil) != tc.wantErr {
+				t.Fatalf("got err %v, want error %v", gotErr, tc.wantErr)
+			}
+			if gotErr != nil {
+				return
+			}
+			want, err := url.Parse(tc.url)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if diff := cmp.Diff(gotSub, want); diff != "" {
```

Comment:
It seems as if you're testing `url.Parse`. Can you explain what's happening here?

Reply:
It's just verifying that the registered handler gets the right URL. Some of the tests are probably overkill since both the real code and the test use `URL.Parse`, but some of them also test other things like an unregistered scheme or a handler returning an error.